File: frontend-shared/types/group.ts
/**
 * Group and community types.
 * These match the shapes the Go backend sends and receives for group-related records.
 * Groups sit between friends and public visibility — members can see catches shared
 * to a group even if they are not personal friends with the person who caught it.
 */

import type { Catch } from './catch';

// The role a member holds within a group.
// Matches the GroupRole constants in backend/models/group.go.
export enum GroupRole {
  Admin = 'admin',
  Moderator = 'moderator',
  Member = 'member',
}

// A community group that users can create, join, and share catches to.
// Mirrors the Go Group struct in backend/models/group.go.
export interface Group {
  id: number;
  name: string;
  description: string;
  created_by: number;
  is_private: boolean;
  // Only present when the current user is an admin and an invite code exists.
  invite_code?: string;
  member_count: number;
  created_at: string; // ISO 8601
  updated_at: string;
}

// A single member of a group with their assigned role.
// Mirrors the Go GroupMember struct in backend/models/group.go.
export interface GroupMember {
  group_id: number;
  user_id: number;
  // The member's username, joined from the users table for display purposes.
  username: string;
  role: GroupRole;
  joined_at: string; // ISO 8601
}

// A pending request from a user to join a private group.
// Mirrors the Go JoinRequest struct in backend/models/group.go.
export interface JoinRequest {
  id: number;
  group_id: number;
  user_id: number;
  // The requester's username, included so admins can review without extra lookups.
  username: string;
  // "pending" | "accepted" | "declined"
  status: string;
  requested_at: string; // ISO 8601
}

// What you send when creating a new group.
export interface CreateGroupRequest {
  name: string;
  description: string;
  is_private: boolean;
}

// What you send when updating an existing group's settings.
// All fields are optional — only the provided fields are updated.
export interface UpdateGroupRequest {
  name?: string;
  description?: string;
  is_private?: boolean;
}

// The paginated response for the group catch feed.
// Catches are returned in the same shape as the regular catch list.
export interface GroupFeedResponse {
  catches: Catch[];
  total: number;
  page: number;
  page_size: number;
}

// An entry in the most-active-members list within GroupStats.
export interface ActiveMember {
  user_id: number;
  username: string;
  catch_count: number;
}

// A single data point in the catch-rate-over-time series within GroupStats.
export interface CatchRateDataPoint {
  // The date bucket (e.g. "2025-03-15" for daily, or "2025-W11" for weekly).
  date: string;
  catch_count: number;
}

// Aggregate statistics for a group.
// Returned by the GetGroupStats endpoint.
export interface GroupStats {
  // Total number of catches shared to this group across all time.
  total_catches: number;
  // The five most-caught species, ranked by frequency.
  top_species: Array<{ species: string; count: number }>;
  // The five most active members by catch count.
  most_active_members: ActiveMember[];
  // Daily or weekly catch counts over the past 30 days.
  catch_rate_over_time: CatchRateDataPoint[];
}

File: backend/models/group.go
package models

import "time"

// GroupRole defines the level of authority a member has within a group.
// Higher roles can manage membership and moderate content.
type GroupRole string

const (
	// GroupRoleAdmin is the creator's role and the highest level of control.
	// Admins can edit group settings, manage all members, generate invite codes,
	// and delete the group entirely.
	GroupRoleAdmin GroupRole = "admin"

	// GroupRoleModerator is a trusted member elevated by an admin.
	// Moderators can approve or decline join requests and remove regular members.
	GroupRoleModerator GroupRole = "moderator"

	// GroupRoleMember is the default role assigned to anyone who joins the group.
	// Members can see the group feed and share catches to the group.
	GroupRoleMember GroupRole = "member"
)

// Group represents a community of anglers who share catches with each other.
// Groups add a visibility layer between friends and public — you can share a
// catch to one or more groups without making it visible to the whole world.
type Group struct {
	// A unique number that identifies this group.
	ID int64 `db:"id" json:"id"`

	// The display name of the group (e.g. "Lake Erie Bass Crew").
	Name string `db:"name" json:"name"`

	// A short description of what the group is about or who it's for.
	Description string `db:"description" json:"description"`

	// The user ID of the member who created the group.
	// That person is automatically given the admin role.
	CreatedBy int64 `db:"created_by" json:"created_by"`

	// When true, the group does not appear in public search results and
	// new members must either use an invite code or have a join request approved.
	IsPrivate bool `db:"is_private" json:"is_private"`

	// A short random code that lets anyone join a private group without
	// going through the approval flow. Only admins can generate or reset it.
	// Null when no invite code has been generated yet.
	InviteCode *string `db:"invite_code" json:"invite_code,omitempty"`

	// The current number of accepted members. Kept as a denormalized count
	// so group cards can be displayed without an extra JOIN.
	MemberCount int `db:"member_count" json:"member_count"`

	CreatedAt time.Time `db:"created_at" json:"created_at"`
	UpdatedAt time.Time `db:"updated_at" json:"updated_at"`
}

// GroupMember records a user's current membership in a group, including their role.
// A row only exists here once the membership is active — pending join requests live
// in the JoinRequest table instead.
type GroupMember struct {
	// The group this membership belongs to.
	GroupID int64 `db:"group_id" json:"group_id"`

	// The user who is a member.
	UserID int64 `db:"user_id" json:"user_id"`

	// The member's current role within the group.
	Role GroupRole `db:"role" json:"role"`

	// When this person joined (or was accepted) into the group.
	JoinedAt time.Time `db:"joined_at" json:"joined_at"`
}

// JoinRequest is created when a user wants to join a private group and
// no invite code is available. An admin or moderator must approve or decline it.
type JoinRequest struct {
	// A unique number that identifies this request.
	ID int64 `db:"id" json:"id"`

	// The private group the user is asking to join.
	GroupID int64 `db:"group_id" json:"group_id"`

	// The user who submitted the request.
	UserID int64 `db:"user_id" json:"user_id"`

	// The current lifecycle state: "pending", "accepted", or "declined".
	Status string `db:"status" json:"status"`

	// When the request was originally submitted.
	RequestedAt time.Time `db:"requested_at" json:"requested_at"`
}
